### PR TITLE
Add LibraryListing functionality

### PR DESCRIPTION
**Pull request by gabemart**
#7

### CODE REVIEW

File: client/src/components/library/LibraryDisplayMenu.jsx [hunk 1]
```diff
@@ -26,6 +33,8 @@ export default class LibraryDisplayMenu extends Component {
 LibraryDisplayMenu.propTypes = {
   sortOrder: PropTypes.oneOf(['last_modified', 'created', 'name']).isRequired,
   onChangeSortOrder: PropTypes.func,
+  reverseSort: PropTypes.bool.isRequired,
+  onChangeReverseSort: PropTypes.func,
```
Comment (jonase) [inline client/src/components/library/LibraryDisplayMenu.jsx]: I suppose we should add `isRequired` to the `func` props as well?

File: client/src/components/library/LibraryListing.jsx [hunk 2]
```diff
@@ -1,7 +1,143 @@
...
+      let dateA = a.created;
+      let dateB = b.created;
+
+      if (this.props.sortOrder === 'last_modified') {
+        dateA = a.modified || dateA;
+        dateB = b.modified || dateB;
+      }
+
+      if (this.props.reverseSort) {
+        output = new Date(a.created) - new Date(b.created);
```
Comment (jonase) [inline client/src/components/library/LibraryListing.jsx]: Should we be using `dateA` & `dateB` here?

File: client/src/components/library/LibraryListingGroup.jsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+export default class LibraryListingGroup extends Component {
+  render() {
+    const listGroupDate = new Date(Date.parse(this.props.listGroup.listGroupDate));
+    const listGroupTitle = listGroupDate.toUTCString(); // TODO: format this
+
+    return (
+      <div className="LibraryListingGroup">
+        {this.props.isSortDateType &&
+          <h3>{listGroupTitle}</h3>
```
Comment (jonase) [inline client/src/components/library/LibraryListingGroup.jsx]: If sorted & grouped by name, do we want to show `A` for the "a" grouping etc.?

**Comment by jonase** [comment]
I agree we could try to refactor the `LibraryListing.render()` method. Would it be possible to create three functions:

* filterEntities(library, filterBy, searchString),
* sortEntities(entities, sortOrder, isReversed),
* groupEntities(entities,  sortOrder)

and in the `LibraryListing.render` basically only do

```javascript
const entities = filterEntities(library, filterBy, searchString);
const sortedEntities = sortEntities(entities, sortOrder, isReversed);
const groupedEntities = groupEntities(sortedEntities, sortOrder);

return (
  <div className={`LibraryListing ${this.props.displayMode}`}>
    <ul>
      ...
);
```

All of this assumes that we have fetched all the entities from the server which is fine for now but something we probably have to revisit later.

**Comment by jonase** [comment]
Another thing I thought about. It's not really the job of LibraryListing to `sort` & `filter` the entities. This could be done in the `LibraryListingGroup` as well.

* `LibraryListing` groups the entities
* `LibraryListingGroup` sorts & filters the entities

This (or similar) approach could perhaps simplify the implementation?

**Comment by gabemart** [comment]
>Another thing I thought about. It's not really the job of LibraryListing to sort the entities. This could be done in the LibraryListingGroup as well.

I agree it makes sense for each `LibraryListingGroup` to sort it's own internal entities. However, I think the LibraryListing will still have to sort the collection of `LibraryListingGroups` itself

**Comment by jonase** [comment]
:+1:

**Pull request merged**